File: cs/3_6_1_observer.cs
using System;
using System.Collections.Generic;

namespace Behavioral.Observer
{
    /// <summary>Інтерфейс спостерігача</summary>
    interface ICustomObserver
    {
        /// <summary>Реакція спостерігача на зміну стану субєкта </summary>
        /// <param name="subject">Субєкт, за яким ведеться спостереження</param>
        void Update(ISubject subject);
    }

    /// <summary>Інтерфейс субєкта, за яким можна вести спостереження</summary>
    interface ISubject
    {
        /// <summary>підписати спостерігача</summary>
        void Attach(ICustomObserver observer);
        /// <summary>відписати  спостерігача</summary>
        void Detach(ICustomObserver observer);
        /// <summary>повідомити всіх підписаних спостерігачів</summary>
        void Notify();
        /// <value>Гетерт для того, щоб спостерігач міг отримати стан субєкта</value>
        int State { get; }
    }

    class Subject : ISubject
    {
        /// <summary> Сатн субєкта визначаємо цілим числом. Стан повинен бути публічно доступний для спостерігача</summary>
        private int _state = 0;

        /// <value> При зміні стану повідомляємо спостерігачів </value>
        public int State
        {
            get
            {
                return _state;
            }
            set
            {
                _state = value;
                Notify();
            }
        }
        private List<ICustomObserver> Observers = new List<ICustomObserver>();
        public void Attach(ICustomObserver observer)
        {
            Observers.Add(observer);
        }
        public void Detach(ICustomObserver observer)
        {
            Observers.Remove(observer);
        }
        public void Notify()
        {
            foreach (var observer in Observers)
            {
                observer.Update(this);
            }
        }
        /// <summary>Змінює стан обєкту на випадкове ціле число від 0 до 9</summary>
        public void GenerateRandomState()
        {
            State = new Random().Next(0, 10);
        }
    }

    /// <summary>Спостерігач виводить в консоль кожну зміну стану</summary>
    class ConsoleLogObserver : ICustomObserver
    {
        public void Update(ISubject subject)
        {
            Console.WriteLine($"Нове значення стану: {subject.State}");
        }
    }

    /// <summary>Спостерігач, який виводить повідомлення про парний стан</summary>
    class EvenObserver : ICustomObserver
    {
        public void Update(ISubject subject)
        {
            if (subject.State % 2 == 0)
            {
                Console.WriteLine("Парне значення стану");
            }
        }
    }
    /// <summary>Підраховує кількість станів, що задовільняють певній умові</summary>
    class CounterObserver : ICustomObserver
    {

        private int Count = 0;
        private Predicate<int> Condition;

        /// <summary>Передаємо умову як предикат</summary>
        public CounterObserver(Predicate<int> Condition)
        {
            this.Condition = Condition;
        }
        public void Update(ISubject subject)
        {
            if (Condition(subject.State))
            {
                Count++;
                Console.WriteLine($"Стан задовільнив умову {Count} разів");
            }
        }
    }

}

File: python/patterns/_3_6_1_observer.py
from __future__ import annotations
from typing import Protocol, Callable
from random import randint


class IObserver [T](Protocol):
    def update(self, new_value: T) -> None: ...


class IObservable [T](Protocol):
    def attach(self, observer: IObserver[T]) -> None: ...
    def detach(self, observer: IObserver[T]) -> None: ...
    def notify(self) -> None: ...


class Subject:  # (IObservable[int])
    def __init__(self) -> None:
        self._state: int = 0
        self._observers: list[IObserver[int]] = []

    @property
    def state(self) -> int:
        return self._state

    @state.setter
    def state(self, value: int) -> None:
        self._state = value
        self.notify()

    def set_random_state(self) -> None:
        self.state = randint(0, 10)

    def attach(self, observer: IObserver[int]) -> None:
        self._observers.append(observer)

    def detach(self, observer: IObserver[int]) -> None:
        if observer in self._observers:
            self._observers.remove(observer)

    def notify(self) -> None:
        for observer in self._observers:
            observer.update(self._state)


class ConsoleLogObserver:  # IObserver[int]
    def update(self, new_value: int) -> None:
        print(f"New value: {new_value}")


class EvenObserver:  # IObserver[int]
    def update(self, new_value: int) -> None:
        if new_value % 2 == 0:
            print(f"New even value: {new_value}")


type Predicate = Callable[[int], bool]


class CounterObserver:  # IObserver[int]
    def __init__(self, condition: Predicate) -> None:
        self.count: int = 0
        self.condition: Predicate = condition

    def update(self, new_value: int) -> None:
        if self.condition(new_value):
            self.count += 1
            print(f"satisfying the condition {self.count} times")
